### PR TITLE
Speed up map processing functions

### PR DESCRIPTION
**Pull request by idshibanov**
Common code was very slow especially on debug build. Big improvement to AI turn speed after recent changes.

### CODE REVIEW

**Comment by ihhub** [comment]
My favorite part of the code :)

**Review by ihhub — COMMENTED** [review]
Hi @idshibanov , I left few comments. Could you please take a look?

File: src/fheroes2/maps/maps.cpp [hunk 1]
```diff
@@ -69,33 +69,29 @@ void Maps::IndexesDistance::Assign( s32 from, const Indexes & indexes, int sort
...
-{
-    while ( objs && *objs ) {
-        if ( *objs == world.GetTiles( index ).GetObject() )
-            return true;
-        ++objs;
+Maps::Indexes MapsIndexesFilteredObjects( const Maps::Indexes & indexes, const u8 * objs, bool checkUnderHero = false )
+{
+    Maps::Indexes result;
+    for ( size_t idx = 0; idx < indexes.size(); ++idx ) {
+        const int objectID = world.GetTiles( indexes[idx] ).GetObject( !checkUnderHero );
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/fheroes2/maps/maps.cpp:76]: Looking at compiled code this is actually 4 extra instructions in assembler (mov, test, jne/jmp, mov 1/0). Enlightening.

File: src/fheroes2/maps/maps.cpp [hunk 2]
```diff
@@ -276,10 +272,11 @@ s32 Maps::GetIndexFromAbsPoint( s32 px, s32 py )
 Maps::Indexes Maps::GetAllIndexes( void )
 {
     Indexes result;
-    result.assign( world.w() * world.h(), 0 );
+    const size_t mapSize = world.getSize();
+    result.reserve( mapSize );
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/fheroes2/maps/maps.cpp:276]: Oh, I forgot about all little tools C++11 brings! This is not only shorter but faster.

File: src/fheroes2/maps/maps.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/fheroes2/maps/maps.cpp:76]: I renamed the both variables to `ignoreHeroes` as flipped the usage.

File: src/fheroes2/maps/maps.cpp [hunk 3]
```diff
@@ -69,33 +69,29 @@ void Maps::IndexesDistance::Assign( s32 from, const Indexes & indexes, int sort
...
-    while ( objs && *objs ) {
-        if ( *objs == world.GetTiles( index ).GetObject() )
-            return true;
-        ++objs;
+Maps::Indexes MapsIndexesFilteredObjects( const Maps::Indexes & indexes, const u8 * objs, bool checkUnderHero = false )
+{
+    Maps::Indexes result;
+    for ( size_t idx = 0; idx < indexes.size(); ++idx ) {
+        const int objectID = world.GetTiles( indexes[idx] ).GetObject( !checkUnderHero );
+        while ( objs && *objs ) {
```
[the remark this answers is not in the document]
Comment (idshibanov) [reply src/fheroes2/maps/maps.cpp:77]: Yes, it is NULL terminated. It was used extensively in the old AI code, but looks like only place this is left is world.cpp. Example:
```
const u8 objs[] = {MP2::OBJ_MONSTER, MP2::OBJ_HEROES, MP2::OBJ_CASTLE, MP2::OBJN_CASTLE, 0};
const MapsIndexes & objv = Maps::GetObjectsPositions( objs );
```

**Comment by ihhub** [comment]
Well done, @idshibanov !

**Pull request merged**